op-e2e/actions/proofs/env.go

```diff
@@ -0,0 +1,181 @@
+	batcher := actions.NewL2Batcher(log, sd.RollupCfg, batcherCfg, sequencer.RollupClient(), miner.EthClient(), engine.EthClient(), engCl)
+
+	addresses := e2eutils.CollectAddresses(sd, dp)
+	cl := engine.EthClient()
+	l2UserEnv := &actions.BasicUserEnv[*actions.L2Bindings]{
+		EthCl:          cl,
+		Signer:         types.LatestSigner(sd.L2Cfg.Config),
+		AddressCorpora: addresses,
+		Bindings:       actions.NewL2Bindings(t, cl, engine.GethClient()),
+	}
+	alice := actions.NewCrossLayerUser(log, dp.Secrets.Alice, rand.New(rand.NewSource(0xa57b)))
+	alice.L2.SetUserEnv(l2UserEnv)
+
+	return &L2FaultProofEnv{
+		log:       log,
+		batcher:   batcher,
+		sequencer: sequencer,
+		engine:    engine,
+		engCl:     engCl,
+		sd:        sd,
+		dp:        dp,
+		miner:     miner,
+		alice:     alice,
+	}
+}
+
+type TestParam func(p *e2eutils.TestParams)
+
+func NewTestParams(params ...TestParam) *e2eutils.TestParams {
+	dfault := actions.DefaultRollupTestParams
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type DeployParam func(p *e2eutils.DeployParams)
+
+func NewDeployParams(t actions.Testing, params ...DeployParam) *e2eutils.DeployParams {
+	dfault := e2eutils.MakeDeployParams(t, NewTestParams())
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type BatcherCfgParam func(c *actions.BatcherCfg)
+
+func NewBatcherCfg(params ...BatcherCfgParam) *actions.BatcherCfg {
+	dfault := &actions.BatcherCfg{
+		MinL1TxSize:          0,
+		MaxL1TxSize:          128_000,
+		DataAvailabilityType: batcherFlags.BlobsType,
+	}
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type OpProgramCfgParam func(p *config.Config)
+
+func NewOpProgramCfg(
+	t actions.Testing,
+	env *L2FaultProofEnv,
+	fi *FixtureInputs,
+	params ...OpProgramCfgParam,
+) *config.Config {
+	dfault := config.NewConfig(env.sd.RollupCfg, env.sd.L2Cfg.Config, fi.L1Head, fi.L2Head, fi.L2OutputRoot, fi.L2Claim, fi.L2BlockNumber)
+
+	// Set up in-process L1 sources
+	dfault.L1ProcessSource = env.miner.L1Client(t, env.sd.RollupCfg)
+	dfault.L1BeaconProcessSource = env.miner.BlobStore()
+
+	// Set up in-process L2 source
+	l2ClCfg := sources.L2ClientDefaultConfig(env.sd.RollupCfg, true)
+	l2RPC := env.engine.RPCClient()
+	l2Client, err := host.NewL2Client(l2RPC, env.log, nil, &host.L2ClientConfig{L2ClientConfig: l2ClCfg, L2Head: fi.L2Head})
+	require.NoError(t, err, "failed to create L2 client")
+	l2DebugCl := &host.L2Source{L2Client: l2Client, DebugClient: sources.NewDebugClient(l2RPC.CallContext)}
+	dfault.L2ProcessSource = l2DebugCl
+
+	if dumpFixtures {
+		dfault.DataDir = t.TempDir()
+		dfault.DataFormat = hostTypes.DataFormatPebble
+	}
+
+	for _, apply := range params {
+		apply(dfault)
+	}
+	return dfault
+}
+
+type FixtureInputParam func(f *FixtureInputs)
+
+func RunFaultProofProgram(t actions.Testing, gt *testing.T, env *L2FaultProofEnv, l2ClaimBlockNum uint64, fixtureInputParams ...FixtureInputParam) error {
```

Comment:
What do you think about making this a method on the `*L2FaultProofEnv` itself?  I kind of lean towards it but a bit on the fence.  I suspect we'll tweak the details of the abstractions we provide a bit as we learn so don't have to get it perfect to start with.

Reply:
Yeah good shout, moved it onto there. Makes sense, otherwise the env is just a container for all the action test backend types.